File: python/day15/part1.py
#!/usr/bin/env python3
import re


class Sensor:
    def __init__(self, sx, sy, bx, by):
        self.x = sx
        self.y = sy
        self.beacon = (bx, by)
        self.distance = abs(sx - bx) + abs(sy - by)


def parse(input_path):
    with open(input_path, 'r') as f:
        return [Sensor(*tuple(map(int, re.findall(r'[-\d]+', line)))) for line in f.readlines()]


def merge_intervals(intervals):
    assert len(intervals) > 0
    intervals.sort()
    res = [intervals[0]]
    for interval in intervals[1:]:
        if res[-1][0] <= interval[0] <= res[-1][-1] + 1:
            res[-1][-1] = max(res[-1][-1], interval[-1])
        else:
            res.append(interval)
    return res


def solve_part1(sensors, target_y):
    intervals = []
    for sensor in sensors:
        half_span = sensor.distance - abs(sensor.y - target_y)
        if half_span <= 0:
            continue  # no covered area in row target_y for this sensor
        intervals.append([sensor.x - half_span, sensor.x + half_span])
    intervals = merge_intervals(intervals)

    beacons = set([s.beacon for s in sensors
                   if s.beacon[1] == target_y
                   and any([a <= s.beacon[0] <= b for [a, b] in intervals])])

    return sum([s[1] - s[0] + 1 for s in intervals]) - len(beacons)


if __name__ == "__main__":
    parsed = parse("data.txt")
    print('Part 1 :', solve_part1(parsed, 2000000))

File: c++/day15/part1.cpp
#include "part1.h"

using namespace std;


namespace Day15 {

    vector<Sensor> parse(const string& filename) {
        vector<Sensor> sensors;
        const regex pattern { "Sensor at x=(-?\\d+), y=(-?\\d+): closest beacon is at x=(-?\\d+), y=(-?\\d+)" };
        for (const auto line : getFileLines(filename)) {
            smatch match;
            regex_search(line.begin(), line.end(), match, pattern);
            Coord position { stoi(match[1]), stoi(match[2]) };
            Coord beacon   { stoi(match[3]), stoi(match[4]) };
            sensors.emplace_back(position, beacon);
        }
        return sensors;
    }

    namespace Part1 {

        vector<pair<int, int>> mergeIntervals(const vector<pair<int, int>>& intervals) {
            if (intervals.size() < 2) {
                return intervals;
            }
            vector<pair<int, int>> merged {};
            pair<int, int> currInterval = intervals[0];
            for (int i = 1; i < intervals.size(); ++i) {
                if (intervals[i].first <= currInterval.second) {
                    // intervals can be merged
                    currInterval.second = max(currInterval.second, intervals[i].second);
                } else {
                    // interval starts after the end of the previous one, create a new interval
                    merged.push_back(currInterval);
                    currInterval = intervals[i];
                }
            }
            merged.push_back(currInterval);
            return merged;
        }

        vector<pair<int, int>> findCoverageForY(const vector<Sensor>& sensors, const int targetY, const int maxBound) {
            vector<pair<int, int>> intervals {};
            for (const auto& sensor : sensors) {
                const int range = Coord::manhattan(sensor.pos, sensor.beacon);
                const int distanceToTargetY = abs(sensor.pos.y - targetY);
                if (range < distanceToTargetY) {
                    continue;
                } else {
                    int left  = sensor.pos.x - (range - distanceToTargetY);
                    int right = sensor.pos.x + (range - distanceToTargetY);
                    if (maxBound != -1) {
                        left  = max(left,  0);
                        right = min(right, maxBound);
                    }
                    intervals.emplace_back(left, right);
                }
            }
            sort(intervals.begin(), intervals.end());
            return mergeIntervals(intervals);
        }

        int solve(const vector<Sensor>& sensors, const int targetY) {
            // get the X of all cells covered by the sensors for that Y
            const auto intervals = findCoverageForY(sensors, targetY);

            // count the number of cells covered by the sensors on that Y
            int coveredCoords = 0;
            for_each (intervals.begin(), intervals.end(), [&](const auto& interval) {
                coveredCoords += interval.second - interval.first + 1;
            });

            // subtract from the covered area the known beacons
            set<int> processedBeaconsX {};
            for (const auto& sensor : sensors) {
                if (sensor.beacon.y != targetY) {
                    // beacon with a different value of Y
                    continue;
                }
                if (processedBeaconsX.find(sensor.beacon.x) != processedBeaconsX.end()) {
                    //beacon already subtracted for a previous sensor
                    continue;
                }
                processedBeaconsX.insert(sensor.beacon.x);
                if (any_of(intervals.begin(),
                           intervals.end(),
                           [&](const auto& interval) { return sensor.beacon.x >= interval.first && sensor.beacon.x <= interval.second; })) {
                    --coveredCoords;
                }
            }

            return coveredCoords;
        }

    }
}
